### ISSUE TITLE
'mnc2nii' corrupts qform, but SCT recommends '-set-sform-to-qform'

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#3249]
more to come soon-- @mpompolas can chip in to document this issue

**Comment by mpompolas** [comment]
I'm dealing with some MINC files.

I converted them to `NIFTI` files with `MINCTOOLS` with two different versions:
- Release 1.9.18 (local computer)
- 0.3.16 (remote server)

Both versions create the issue below:

`sct_label_vertebrae` throws the sform-qform inconsistency
`ValueError: Image sform does not match qform
`

I follow the suggested solution to assign the same value to both:
`sct_image -i "$NIFTIFILE" -set-sform-to-qform -o "$NIFTIFILE"`

However when I perform `sct_deepseg_sc` on the modified `NIFTI`, the spinalcord segmentation and the modified NIFTI are not corregistering because the headers are assigning them to different spaces

**Comment by joshuacwnewton** [comment]
Linking for relevance: 
* #2429 was where sform/qform mismatches were first brought up
* #2858 was where the raised error + suggested solution were implemented

**Comment by joshuacwnewton** [comment]
@mpompolas Are you able to share the images you're working with?

**Comment by mpompolas** [comment]
sorry, it is private. I'll check if this can be reproduced from any minc file

**Comment by joshuacwnewton** [comment]
No worries, very understandable. Thank you for checking. :slightly_smiling_face:

**Comment by mpompolas** [comment]
I was able to reproduce the error on a random nifti file I chose from spinegeneric dataset (`sub-juntendo750w01_T1w.nii.gz`).

Steps to follow:
1. install minctools (used Release 1.9.18 ): https://github.com/BIC-MNI/minc-toolkit-v2#installation
2. convert nifti to minc: `nii2mnc sub-juntendo750w01_T1w.nii.gz test.mnc.gz`
3. convert back to nifti: `mnc2nii test.mnc.gz test.nii && gzip test.nii`
4. `sct_deepseg_sc -i test.nii.gz -c t1`   (the registration between test.nii.gz and test_seg.nii.gz is correct at this point)
5. `sct_label_vertebrae -i test.nii.gz -c t1 -s test_seg.nii.gz`  (this should fail due to ValueError: Image sform does not match qform)
6. `sct_image -i test.nii.gz -set-sform-to-qform`  (The proposed solution is to use sct_image with the flag -set-sform-to-qform)
7. `sct_deepseg_sc -i test.nii.gz -c t1`  (however if I redo the segmentation, the new: test.nii.gz and test_seg.nii.gz are not coregistered anymore due to the header corruption from mnc2nii)

In short: the usage of sct_image -set-sform-to-qform in step 6 is incompatible with the header changes that mnc2nii makes on step 3.

**Comment by joshuacwnewton** [comment]
Thank you so much for taking the time to dig deeper and to create reproduction steps, @mpompolas. :slightly_smiling_face: 

> (however in this segmentation, the new: test.nii.gz and test_seg.nii.gz are not coregistered anymore due to the header "corruption" from mnc2nii)

When you say 'corruption', are you suggesting that `mnc2nii` has an upstream issue? In other words, is this problem actually 2 separate issues?

1. `mnc2nii` produces incorrect headers
2. SCT then recommends a poor solution that doesn't address the issue

If so, then maybe the next step should be to clearly document how `mnc2nii` is corrupting the headers.

**Comment by joshuacwnewton** [comment]
I followed the reproduction steps above, then compared the headers before/after conversion:

```console
> nii2mnc t1w.nii.gz t1w.mnc.gz
> mnc2nii t1w.mnc.gz t1w-converted.nii && gzip t1w-converted.nii
> fslhd t1w.nii.gz | xclip -sel c
> fslhd t1w-converted.nii.gz | xclip -sel c
```

Here's the full diff between image headers: https://www.diffchecker.com/1NYzr6HV

The `sform` is actually OK here -- it's the `qform` that was corrupted and needs fixing:

```python
# Before conversion
qform_name	Scanner Anat
qform_code	1
qto_xyz:1	1.000000 0.000000 0.000000 -31.219284 
qto_xyz:2	0.000000 1.000000 0.000000 -36.213394 
qto_xyz:3	0.000000 0.000000 1.000000 -21.321671 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 

# After conversion
qform_name	Unknown
qform_code	0
qto_xyz:1	1.000000 0.000000 0.000000 0.000000 
qto_xyz:2	0.000000 1.000000 0.000000 0.000000 
qto_xyz:3	0.000000 0.000000 1.000000 0.000000 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
```

So, when we recommend  `-set-sform-to-qform`, we're doing the opposite of what we should be doing!

Some thoughts:

* For detecting this problem, we could check for sforms/qforms that are identity matrices. 
   * From what I understand of [coordinate systems](https://nipy.org/nibabel/coordinate_systems.html#voxel-coordinates-and-points-in-space), typical images shouldn't have affine matrices like this, right? (It seems unlikely that the FOV center would be exactly on the magnet isocenter with no rotation/translation/scaling.)
   * So, perhaps we could we interpret an identity matrix as an "uninitialized" affine? (Much like we do matrices that are all zeros.)
* For fixing this problem, we should expand `-set-sform-to-qform` so that it can be performed in either direction, now that we know there's a valid use case for setting the qform to the sform.

**Comment by joshuacwnewton** [comment]
Ah, I've just looked at the pages for [`mnc2nii`](http://bic-mni.github.io/man-pages/man/mnc2nii.html) and  [`nii2mnc`](http://bic-mni.github.io/man-pages/man/nii2mnc) and they state: 

> **KNOWN BUGS**
>
> Current handling of NIfTI-1 qform and sform coordinate transforms should probably be revised as the NIfTI group clarifies the correct usage of these fields.

**Comment by jcohenadad** [comment]
> For detecting this problem, we could check for sforms/qforms that are identity matrices.
> - From what I understand of coordinate systems, typical images shouldn't have affine matrices like this, right? (It seems unlikely that the FOV center would be exactly on the magnet isocenter with no rotation/translation/scaling.)
> - So, perhaps we could we interpret an identity matrix as an "uninitialized" affine? (Much like we do matrices that are all zeros.)
> - This ties into other checks I'd like to run as part of #3283 (comment).

I am not a big fan of this strategy. Two examples where this would be problematic:
- template images (e.g. MNI template) sometimes have identity matrix with zeroed vector for the 4th column. They are not _wrong_ per se, but they are just defined in a "new" space coordinate that is different from that of the scanner. Hence the introduction of qform_code 3 and 4 (Talairach and MNI space).
- Sometimes users are doing experiments where the absolute qform does not matter. So they would do a bunch of processing on the image array only (not caring about the header), usually on Matlab, and then save the image with a "virgin" qform header. We don't want SCT to _not_ process these images.

>> KNOWN BUGS
>> Current handling of NIfTI-1 qform and sform coordinate transforms should probably be revised as the NIfTI group clarifies the correct usage of these fields.

I'm a bit suspicious about this statement. First of all, the page indicates "copyright 2005" at the bottom and i'm not sure when was the last time it was updated. Secondly, the "NIfTI group", which happens to be the vast majority of neuroimaging users on the planet, did came up with a pretty clear usage of these fields. The problem really comes from the various software that use those fields in a different way (e.g. SPM/FSL vs. ITK/ANTs). More context in https://github.com/neuropoly/spinalcordtoolbox/issues/3283.

**Comment by joshuacwnewton** [comment]
>  Two examples where this would be problematic:

Thanks for pointing this out! I was wondering if there were valid use-cases for affines with an identity matrix, so I appreciate the extra context.

In that case, perhaps we should edit [our warning message](https://github.com/neuropoly/spinalcordtoolbox/blob/6b32309b9a647b0755fc931bd5f8319493a01218/spinalcordtoolbox/image.py#L307-L309) like this:

```diff
                logger.error("Image {} has different qform and sform matrices. This can produce incorrect results. "
-                             "Consider setting the two matrices to be equal by running:\n"
-                             "sct_image -i {} -set-sform-to-qform -o {}.".format(
+                             "Please use 'sct_image -i {} -hd' to check the header for issues (such as an "
+                             "uninitialized matrix). Then, consider running either 'sct_image -set-sform-to-qform' "
+                             "or 'sct_image -set-qform-to-sform' to fix any issues.".format(
```

This way, we're encouraging users o be aware of the contents of their headers, rather than following a suggested fix without considering why.

Note: I've written `sct_image -hd` here, which is just a flag to display the header, as requested in #540. (Another good reason why we should implement this. I'd like to get started on that ASAP.)

**Comment by joshuacwnewton** [comment]
> I'm a bit suspicious about this statement. First of all, the page indicates "copyright 2005" at the bottom and i'm not sure when was the last time it was updated. Secondly, the "NIfTI group", which happens to be the vast majority of neuroimaging users on the planet, did came up with a pretty clear usage of these fields. The problem really comes from the various software that use those fields in a different way (e.g. SPM/FSL vs. ITK/ANTs). More context in #3283.

Good point. I did a little more digging, and the most relevant discussion I've seen on the topic (on `minctools`'s end) is here: https://github.com/BIC-MNI/minc-tools/issues/42#issuecomment-243794522

> What I meant is this: in general, conversion nii2mnc -> -> mnc2nii should preserve the voxel-to-world transformation. 

So, even in 2016 there was an awareness that these tools do not preserve the affines of Nifti images.
